### PR TITLE
Activate doxygen build on CI

### PR DESCRIPTION
**Pull request by rasky**
Upgrade doxygen-public.conf to the latest version (via doxygen -u)
and turn on warnings as errors, so that the build fails if a PR
introduces undocumented text.

Also go through the whole libdragon and add missing documentation.

### CODE REVIEW

**Comment by anacierdem** [comment]
Will close #77

**Comment by meeq** [comment]
I will take a swing at resolving the outstanding warnings for the Transfer Pak subsystem:
```
libdragon/include/tpak.h:12: warning: Member TPAK_ERROR_INVALID_ARGUMENT (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:13: warning: Member TPAK_ERROR_NO_TPAK (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:14: warning: Member TPAK_ERROR_NO_CONTROLLER (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:15: warning: Member TPAK_ERROR_UNKNOWN_BEHAVIOUR (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:16: warning: Member TPAK_ERROR_NO_CARTRIDGE (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:17: warning: Member TPAK_ERROR_ADDRESS_OVERFLOW (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:19: warning: Member TPAK_STATUS_READY (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:20: warning: Member TPAK_STATUS_WAS_RESET (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:21: warning: Member TPAK_STATUS_IS_RESETTING (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:22: warning: Member TPAK_STATUS_REMOVED (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:23: warning: Member TPAK_STATUS_POWERED (macro definition) of file tpak.h is not documented.
libdragon/include/tpak.h:25: warning: Member gbc_support_type (enumeration) of file tpak.h is not documented.
libdragon/include/tpak.h:32: warning: Compound old_gbc_title is not documented.
libdragon/include/tpak.h:34: warning: Member title[15] (variable) of struct old_gbc_title is not documented.
libdragon/include/tpak.h:35: warning: Member gbc_support (variable) of struct old_gbc_title is not documented.
libdragon/include/tpak.h:38: warning: Compound new_gbc_title is not documented.
libdragon/include/tpak.h:40: warning: Member title[11] (variable) of struct new_gbc_title is not documented.
libdragon/include/tpak.h:41: warning: Member manufacturer_code[4] (variable) of struct new_gbc_title is not documented.
libdragon/include/tpak.h:42: warning: Member gbc_support (variable) of struct new_gbc_title is not documented.
libdragon/include/tpak.h:45: warning: Compound gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:47: warning: Member entry_point[4] (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:48: warning: Member logo[48] (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:54: warning: Member new_licensee_code (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:55: warning: Member is_sgb_supported (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:56: warning: Member cartridge_type (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:57: warning: Member rom_size_code (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:58: warning: Member ram_size_code (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:59: warning: Member destination_code (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:60: warning: Member old_licensee_code (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:61: warning: Member version_number (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:62: warning: Member header_checksum (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:63: warning: Member global_checksum (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:64: warning: Member overflow[16] (variable) of struct gameboy_cartridge_header is not documented.
libdragon/include/tpak.h:72: warning: return value 'bit' of tpak_get_status has multiple documentation sections
libdragon/include/tpak.h:72: warning: return value 'bit' of tpak_get_status has multiple documentation sections
libdragon/include/tpak.h:72: warning: return value 'bit' of tpak_get_status has multiple documentation sections
libdragon/include/tpak.h:72: warning: return value 'bit' of tpak_get_status has multiple documentation sections
libdragon/src/tpak.c:161: warning: return value 'bit' of tpak_get_status has multiple documentation sections
libdragon/src/tpak.c:161: warning: return value 'bit' of tpak_get_status has multiple documentation sections
libdragon/src/tpak.c:161: warning: return value 'bit' of tpak_get_status has multiple documentation sections
libdragon/src/tpak.c:161: warning: return value 'bit' of tpak_get_status has multiple documentation sections
libdragon/src/tpak.c:25: warning: explicit link request to 'tpak_set_power(false)' could not be resolved
libdragon/src/tpak.c:28: warning: Member TPAK_POWER_ON (macro definition) of file tpak.c is not documented.
libdragon/src/tpak.c:29: warning: Member TPAK_POWER_OFF (macro definition) of file tpak.c is not documented.
libdragon/src/tpak.c:30: warning: Member TPAK_POWER_ADDRESS (macro definition) of file tpak.c is not documented.
libdragon/src/tpak.c:31: warning: Member TPAK_BANK_ADDRESS (macro definition) of file tpak.c is not documented.
libdragon/src/tpak.c:32: warning: Member TPAK_STATUS_ADDRESS (macro definition) of file tpak.c is not documented.
libdragon/src/tpak.c:33: warning: Member TPAK_STATUS_CARTRIDGE_ABSENT (macro definition) of file tpak.c is not documented.
libdragon/src/tpak.c:35: warning: Member TPAK_DATA_ADDRESS (macro definition) of file tpak.c is not documented.
libdragon/src/tpak.c:37: warning: Member BLOCK_SIZE (macro definition) of file tpak.c is not documented.
libdragon/src/tpak.c:38: warning: Member BANK_SIZE (macro definition) of file tpak.c is not documented.
```

These warnings also remain:
```
libdragon/include/controller.h:149: warning: Member data (variable) of struct SI_origdat_gc is not documented.
libdragon/include/controller.h:154: warning: Compound SI_origdat_gc is not documented.
libdragon/include/controller.h:156: warning: Member deadzone0 (variable) of struct SI_origdat_gc is not documented.
libdragon/include/controller.h:157: warning: Member deadzone1 (variable) of struct SI_origdat_gc is not documented.
libdragon/include/controller.h:157: warning: Member gc[4] (variable) of struct controller_origin_data is not documented.
libdragon/include/controller.h:171: warning: Compound controller_origin_data is not documented.
```

**Comment by rasky** [comment]
@meeq are you going to handle the controller.h ones as well?

**Comment by meeq** [comment]
@Rasky I am reviewing the controller subsystem routines for GameCube controllers to update the Doxygen comments, but the implementation leaves me scratching my head about why we would support it in its current state.

1. You need to make a passive GCN->N64 controller adapter in order to use it. The use-case for using GameCube Joybus commands on a Nintendo 64 is extremely esoteric and was never supported by Nintendo or any third-party.
2. `controller_read_gc` and `controller_read_gc_origin` are not integrated with the rest of the controller subsystem. A developer would need completely separate input code-paths to support GameCube controllers.
3. `struct controller_data`  is wasteful, confusing, and unnecessary: You will never get N64 and GCN controller data in the same result, so half of the struct is always padding.

I don't necessarily think  we should remove the GC controller commands completely, but I think they need to be isolated from the rest of the controller subsystem and flagged as highly-experimental if it even makes sense to keep it at all.

**Comment by meeq** [comment]
For now, I'll add some documentation to appease the warnings, and I'll open a separate issue for discussing the removal/reworking of GameCube controller support.

**Edit 1:** https://github.com/rasky/libdragon/pull/11
**Edit 2:** https://github.com/DragonMinded/libdragon/issues/218

File: include/ay8910.h [hunk 1]
```diff
@@ -51,46 +51,60 @@ extern "C" {
...
-	uint8_t shape;
+	uint16_t period;           ///< Period (in ticks) of the envelope
+	uint8_t shape;             ///< Shape of the envelope (jigsaw, etc.)
 
-	int16_t step;
-	uint8_t attack;
-	uint8_t alternate, hold, holding;
+	uint8_t attack;            ///< 0x0 if in attack, 0xF if in the release
+	uint8_t alternate;         ///< True if attack and release alternate (jigsaw)
+    uint8_t hold               ///< True if the envelope holds after attack
```
Comment (meeq) [inline include/ay8910.h]: Missing semicolon here

**Comment by rasky** [comment]
@anacierdem @meeq this is ready.

The CI now also builds Doxygen docs, and fails in case of any error (including missing documentation). Thanks to @meeq for helping bringing warnings to zero. The docs are published on GitHub pages (via the `gh_pages` orphan branch), into the `ref` subdirectory. Building is done everywhere, publishing only from branch "trunk".

So on GitHub pages we will get a standard Jekyll site at the root URL (committed as `website/` in git), and then the Doxygen reference in the /ref URL. If we want, we can build a little Jekyll site to describe libdragon. I took the chance to also rewrite `README.md`, which is also published as the only page in the website via Jekyll.

Let me know if this sounds OK to you.

**Comment by sp1187** [comment]
I personally think the history would be a bit cleaner if the CI configuration and the documentation fixes were on separate commits.

**Comment by rasky** [comment]
> I personally think the history would be a bit cleaner if the CI configuration and the documentation fixes were on separate commits.

Agreed, I just did that.

**Comment by anacierdem** [comment]
It sounds ok, is there a way to preview it?

**Comment by rasky** [comment]
Website:
https://rasky.github.io/libdragon

Doxygen:
https://rasky.github.io/libdragon/ref/

**Pull request merged**